### DISCUSSION TITLE
New plugin for drawing on canvas and mouse tracking

### DISCUSSION THREAD

**Discussion by GEJ1** [jspsych/jsPsych#1370] [Show and tell]
Hello Everyone,

I'm working on a new plugin for drawing on canvas and tracking the cursor position.
You can see it an working example running on Cognition [https://nrz7zmmo5m.cognition.run/](https://github.com/GEJ1/draw-and-mouse-tracking) and the code in this repository [https://github.com/GEJ1/draw-and-mouse-tracking]( https://github.com/GEJ1/draw-and-mouse-tracking).

![new_plugin](https://user-images.githubusercontent.com/8774034/104467742-710b8300-5595-11eb-8230-1c84f456dfd9.gif)

Because of the type of experiment I want to do, the plugin work as follows: The image is presented at the beginning of the trial, when you click, it starts to draw and when you release the button the trial ends. But I think that we can made a more flexible plugin.

**Parameters:**
* Images to present
* canvas_background_color
* content_wrapper_color
* Draw line color
* Line width
* Canvas size

**Data Generated:**
*  pos_tracking: The X and Y position of the cursor in the canvas while you are drawing.
*  cursor_time : The time relative to the start of the trial for every (X,Y) position. (_This can be useful to calculate velocity_)
* X_click: The position of the click in X coordinate.
* Y_Click: The position of the click in Y coordinate.
* rt: trial_duration

> EDIT: 
> Solved:
> I'm starting 3 variables in the global scope of the experiment file because I'm not sure how to do it properly inside the plugin:
> let x = 0;
> let y = 0; 
> isDrawing = false;`
> 

I'll be happy if anyone share their critics and comments in order to improve it.

I hope it serves the community and can become a plugin that can be used for different experiments!

**Comment by jodeleeuw** [comment]
This is really neat @GEJ1!

I'm not sure why you need the global variables in the experiment script instead of the plugin file. Is there a reason those variables don't work if you just stick them in the `plugin.trial` function?

**Reply by GEJ1** [reply]
Thank you @jodeleeuw ! 😃

I put it in plugin.trial function and now works. Maybe was a problem in an older version.
